core/lib/external_price_api/src/coingecko_api.rs

```diff
@@ -0,0 +1,129 @@
+use std::{collections::HashMap, str::FromStr};
+
+use async_trait::async_trait;
+use bigdecimal::BigDecimal;
+use chrono::Utc;
+use reqwest;
+use serde::{Deserialize, Serialize};
+use zksync_config::configs::ExternalPriceApiClientConfig;
+use zksync_types::{base_token_ratio::BaseTokenAPIPrice, Address};
+
+use crate::{address_to_string, PriceAPIClient};
+
+#[derive(Debug)]
+pub struct CoinGeckoPriceAPIClient {
+    base_url: url::Url,
+    client: reqwest::Client,
+}
+
+const COINGECKO_AUTH_HEADER: &str = "x-cg-pro-api-key";
+const ETH_ID: &str = "ethereum";
+const USD_ID: &str = "usd";
+
+impl CoinGeckoPriceAPIClient {
+    pub fn new(config: ExternalPriceApiClientConfig) -> Self {
+        let client = if let Some(api_key) = &config.api_key {
+            let mut headers = reqwest::header::HeaderMap::new();
+            headers.insert(
+                reqwest::header::HeaderName::from_static(COINGECKO_AUTH_HEADER),
+                reqwest::header::HeaderValue::from_str(api_key)
+                    .expect("Failed to create header value"),
+            );
+
+            reqwest::Client::builder()
+                .default_headers(headers)
+                .timeout(config.client_timeout())
+                .build()
+                .expect("Failed to build reqwest client")
+        } else {
+            reqwest::Client::new()
+        };
+
+        Self {
+            base_url: config.base_url(),
+            client,
+        }
+    }
+
+    async fn get_token_price_by_address(&self, address: Address) -> anyhow::Result<f64> {
+        let address_str = address_to_string(&address);
+        let price_url = self
+            .base_url
+            .join(
+                format!(
+                    "/api/v3/simple/token_price/ethereum?contract_addresses={}&vs_currencies={}",
+                    address_str, USD_ID
+                )
+                .as_str(),
+            )
+            .expect("failed to join URL path");
+
+        let response = self.client.get(price_url).send().await?;
+        if !response.status().is_success() {
+            return Err(anyhow::anyhow!(
+                "Http error while fetching token price. Status: {}, token_addr: {}, msg: {}",
+                response.status(),
+                address_str,
+                response.text().await.unwrap_or(String::new())
+            ));
+        }
+
+        let cg_response = response.json::<CoinGeckoPriceResponse>().await?;
+        match cg_response.get_price(&address_str, &USD_ID.to_string()) {
+            Some(&price) => Ok(price),
+            None => Err(anyhow::anyhow!(
+                "Price not found for token: {}",
+                address_str
+            )),
+        }
+    }
+
+    async fn get_token_price_by_id(&self, id: String) -> anyhow::Result<f64> {
+        let price_url = self
+            .base_url
+            .join(format!("/api/v3/simple/price?ids={}&vs_currencies={}", id, USD_ID).as_str())
```

Comment:
`vs_currencies` seem to [support eth](https://docs.coingecko.com/reference/simple-supported-currencies), so you don't need two requests; one should be enough.
Otherwise your calculation will not be atomic and it will loose precision.